HDG for advection-diffusion on the sphere

amartinhuertas opened the pull request:
Super WIP. Opening PR in draft mode ... just for discussion purposes

davelee2804 committed:
first commit of the HDG advection branch for GridapGeosciences (using
GridapHybrid and related version of Gridap)

davelee2804 committed:
adding the Manifest.toml so we can build off the right Gridap and
GridapHybrid versions

davelee2804 committed:
[ci skip] adding first commit of the hdg advection driver and solver
(unfinished)

davelee2804 committed:
[ci skip] adding some more detail to the HDG advection test solver

davelee2804 committed:
[ci skip] adding first order time stepping for the hdg advection test

davelee2804 committed:
second order implicit runge-kutta time stepping

davelee2804 committed:
fixing an inital condition bug, and also using the velocity function in
the assembly terms

davelee2804 committed:
[ci skip] fixing some bugs in the advection hdg solver

davelee2804 commented:
Hey @amartinhuertas , I'm having some basic trouble with the HDG advection on the sphere. I am getting the error:
> It is not possible to perform the operation "dot" on the given cell fields.

here:
https://github.com/gridapapps/GridapGeosciences.jl/blob/hdg_adv/src/AdvectionHDG.jl#L19
Currently I am passing in `un` directly from the function:
https://github.com/gridapapps/GridapGeosciences.jl/blob/hdg_adv/driver/sequential/SBRAdvectionHDG.jl#L19
However I have also tried projecting `un` onto a FESpace via
https://github.com/gridapapps/GridapGeosciences.jl/blob/hdg_adv/src/AdvectionHDG.jl#L57
https://github.com/gridapapps/GridapGeosciences.jl/blob/hdg_adv/src/AdvectionHDG.jl#L90
I'm wondering if the vector field on the manifold and the L2 reference element vector field (as I've defined it) are somehow incompatible? (Note that D = 2)...

amartinhuertas commented:
Can you please separate the terms in the bilinear form one by one as in e.g., https://github.com/gridap/GridapHybrid.jl/blob/main/test/DarcyHDGTests.jl#L71
and let me know which is the one(s) that triggers the error? Thanks!

amartinhuertas commented:
> (Note that D = 2).

Did you try with D=3?

davelee2804 commented:
...I tried `D=3`, but it failed even earlier (I can't remember where exactly sorry). Thanks for the suggestion, the failure is at:
`  _op2 = (un⋅n)`

amartinhuertas commented:
can you attach a file with the whole backtrace? I would like to understand the cause behind the inability to perform the dot product operator.

davelee2804 committed:
[ci skip] adding brackets around dot products

davelee2804 commented:
[err.log](https://github.com/gridapapps/GridapGeosciences.jl/files/9016173/err.log)

amartinhuertas commented:
Ok, digging into the error log, I saw this line of code:

> setindex!(A::Vector{Gridap.TensorValues.VectorValue{2, Float64}}, x::Gridap.TensorValues.VectorValue{3, Float64}, i1::Int64)

There we are trying to set a 3-components field vector (`Gridap.TensorValues.VectorValue{3, Float64}`) as an entry of a Vector of 2-component field vectors (`Vector{Gridap.TensorValues.VectorValue{2, Float64}}`). Thus, as you say, it seems we are messing things up with the number of spatial dimensions. 

What is the error that you get with D=3?

davelee2804 commented:
Here is the D=3 log file:
[err-2.log](https://github.com/gridapapps/GridapGeosciences.jl/files/9016220/err-2.log)

amartinhuertas commented:
ok. Use D=3 ONLY for this line, i.e., 

`reffeᵤ = ReferenceFE(lagrangian,VectorValue{3,Float64},order;space=:P)`

The rest untouched.

davelee2804 commented:
...Same error as the original, as far as I can see: :( 
[err-3.log](https://github.com/gridapapps/GridapGeosciences.jl/files/9016283/err-3.log)

davelee2804 commented:
...I just got this error:
[err-4.log](https://github.com/gridapapps/GridapGeosciences.jl/files/9016349/err-4.log)
...with this source:
[AdvectionHDG.log](https://github.com/gridapapps/GridapGeosciences.jl/files/9016350/AdvectionHDG.log)
(so `un` is here projected onto `U`, but is maybe not being interpreted as a vector field)?

amartinhuertas committed:
Fixing push_normal for rectangular jacobians
(see https://github.com/gridapapps/GridapGeosciences.jl/pull/34#issuecomment-1170662488)

davelee2804 committed:
projecting the velocity field onto a L2 vector space and using this in
the HDG advection assembly

amartinhuertas committed:
Merge branch 'refactoring_cubed_sphere_data_structures' of github.com:gridapapps/GridapGeosciences.jl into hdg_adv

Conflicts:
Project.toml

amartinhuertas committed:
Misc fixes to have a running SBRAdvectionHDG.jl (one that does
not crash). The current results it is generating are wrong.

amartinhuertas commented:
> Hey @amartinhuertas , I'm having some basic trouble with the HDG advection on the sphere. I am getting the error:

For the records, this was addressed in https://github.com/gridapapps/GridapGeosciences.jl/pull/34/commits/1586cf7729879573c9598fe9dcf56acfad16db25

amartinhuertas commented:
Hi @davelee2804, 

please read the below carefully, let me know if you have questions, and please answer my questions (no hurries).

Current status of this development:
* I have solved several minor issues in `GridapHybrid.jl` when dealing with manifolds. (see https://github.com/gridap/GridapHybrid.jl/pull/23/files). These have been already merged to `GridapHybrid.jl#main`. As we are pointing there in the `GridapGeosciences.jl#hdg_adv` branch, after pulling from remote `hdg_adv`, you just have to write `up` in the Julia REPL package manager to update `GridapHybrid.jl`.
* Now the HDG driver for advection on the sphere runs bottom-top (it does not crash). **HOWEVER** it does not produce the correct result, in particular, it does not seem to time evolve. 

Important remarks:
* Let us stick (by now) to a 2nd-order polynomial approximation of the sphere (note that in the latest version of the driver I am passing polynomial degree `2` in this call https://github.com/gridapapps/GridapGeosciences.jl/pull/34/files#diff-d6ace757a93640319c464479d34d4617d5d53c419c630a9afee6b9c8db9b8d46R41 ). I still have to go through the analytical mapping cubed sphere mesh to see whether it properly represents the 1D skeleton in 3D ambient space. 
* I had to change the order of the operands in terms like this `m*p*((un⋅nₒ) + τ*(n⋅nₒ))` as `((un⋅nₒ) + τ*(n⋅nₒ))*m*p`. In the former form, `GridapHybrid.jl` generates an internal error (I will solve it when I have time, I think there are some functions missing, but I still have to fully understand this). However, in the latter form, it does not.

Questions:
* Do you have a reference where the formulation you are aiming to implement is written? If not, can you write it? This may help to (1) try to understand what you are trying to do. (2) try to determine if the code is doing what you are trying to do or not.
* I have seen you have terms like `n \cdot n`. Given that n is a unit vector, isnt `n \cdot n` always 1?
* Note that `no` and `n` are defined differently. `n` is the tangent outward unit normal (it always points outwards). `no` points outwards if the corresponding cell is the OWNER of the edge, and points inwards if the corresponding cell is not the owner. Thus, for a given cell K, and edge E in the bounday of K, `n \cdot no` is 1 if K owns E, and it is not 1 (its value depends on the cosine of the angle among the two normals) if it does not own it. We assign the owner cell of an edge arbitrarily to be the cell with the minimum global identifier among the two cells surrounding the edge.

davelee2804 commented:
Thank you @amartinhuertas , that was quick!

I'm not surprised there are still bugs in the code. I update to your current version of GridapHybrid.jl and take a look (hopefully tomorrow).

> Let us stick (by now) to a 2nd-order polynomial approximation of the sphere

Great idea, yes, let's do this

> Do you have a reference where the formulation you are aiming to implement is written?

It is solid body rotation of the flux form advection equation, so the (constant) velocity field is 0 in the north-south direction and `cos(\phi)` in the east-west direction (with `\phi` being the latitude). This is the 'Williamson 1' test case ('Advection of a Cosine Bell) from this paper here: https://www.sciencedirect.com/science/article/pii/S0021999105800166 (with `\alpha=0`). However I am using a Gaussian not a cosine as this is infinitely differentiable, so a little nicer to work with.

> Given that n is a unit vector, isnt n \cdot n always 1

Yes, as you say. I have kept the `n\cdot n` as I wanted to be consistent with respect to the `n\cdot n_o` terms in the jumps, just until I am comfortable with the method, then I will remove

> Note that no and n are defined differently

Yes, I noticed this, thanks. So `n\cdot n_o` represents a jump, ie a `[[ . ]]` operator yes?

davelee2804 committed:
[ci skip] minor tidying up and checking the initial conditions are ok

amartinhuertas commented:
> I'm not surprised there are still bugs in HGD advection the code. I update to your current version of GridapHybrid.jl and take a look (hopefully tomorrow).

Please note that the bugs might not come only from your side. We are testing GridapHybrid with manifolds for the first time and there might be still pitfalls/bugs to be determined/fixed. I think you are aware of this, but just to be 100% sure. I would be extremely careful, and may be try to think in simpler tests we can write to determine whether integration of fields over \partialK is actually working as expected with manifolds.

 > The weak HDG form of the flux form advection equation is based on the formulation in Kang et. al. JCP 2020

Can you send me the particular Eq. number in that paper?  

> Yes, I noticed this, thanks. So n\cdot n_o represents a jump, ie a [[ . ]] operator yes?

Please note that, in general, the jump operator may come in two different flavours:
* `u^+ - u^-`, where u can be a scalar or vector double-valued discontinuous field. If u is scalar, then the jump is scalar, if it is a vector, then it is a vector. This kind of jumps are used in the 
* `u^+ n^+ + u^- n^-`. Where n is the outward unit normal as seen from the perspective of the two cells around.  If u is an scalar, the jump is a vector. If u is a vector, the jump is an scalar. Note that for manifolds n^+ is not equal to -n^- in general.

See def 14.4 and remark 14.5 in 
https://www.math.tamu.edu/~guermond/M610_SPRING_2016/chap14.pdf

n \cdot n0 seems actually another thing (like trying to give a global orientation to the normal, and then adjusting if  the normal from the current cell does not match the one given in the global orientation), but perhaps I am missing something.  I am not saying it is incorrect, I am bit confused actually.

davelee2804 commented:
Hey @amartinhuertas , the relevant equations in Kang et. al. are (3a), (16) and (17).

With respect to the normals we also have that `n^+ = -n^-`, so assuming `K^+` is the cell owning the normal, then `n^+ = n_o` and `n^- = -n_o` so `u\cdot n\_o` integrated over `\partial K` from both sides would be `(n^+ - n^-)\cdot n_o` (ie: a scalar jump). At least that is my understanding. I also get very confused with these binary choices (that is why I left the `n\cdot n` terms in the formulation :) ), there is a very strong chance that I am wrong...

davelee2804 commented:
Hey @amartinhuertas , just a heads up, when I check the `sum()` of the input fields `un`, `pn`, these are non-zero, however the half step field, `ph` has zero `sum()`. I've also looked at individual matrix terms as vectors (assuming pn for the trial function), and these also have non-zero `sum()`

amartinhuertas commented:
> u\cdot n\_o integrated over \partial K from both sides would be (n^+ - n^-)\cdot n_o

Why u\cdot n\_o integrated over both sides and then summed is (n^+ - n^-)\cdot n_o? In the latter expression, I do not see the u. ??? What am I missing ...

> With respect to the normals we also have that n^+ = -n^-

I have doubts in regards to this equivalence for discretized manifolds. I think this equivalence is only approximate in this case, as the normals might be contained in different planes (the ones defined by the cells). At least the code does not assume this, and compute n^+ and n^- from each cell without relying on this equivalence. I guess that the finer the mesh, the more accurate the approximation, but not totally sure. I can write an small numerical experiment to test whether this is true or not.

amartinhuertas commented:
> Hey @amartinhuertas , just a heads up, when I check the sum() of the input fields un, pn, these are non-zero, however the half step field, ph has zero sum(). I've also looked at individual matrix terms as vectors (assuming pn for the trial function), and these also have non-zero sum()

Thanks for reporting this! I found another BUG for GridapHybrid.jl + manifolds. I have fixed it in this PR https://github.com/gridap/GridapHybrid.jl/pull/24/files. I have merged it into `main`.  Please execute `up` in the package manager to get this BUG fix on your env. 

I have executed the HDG advection driver after the bug fix, and it now time evolves, but it does seem to generate just rubbish. We still have to find extra bug fixes, formulation pitfalls etc.

amartinhuertas commented:
@davelee2804  ... for convenience, I wrote a MWE (Minimum Working Example) to find the previous BUG. I am copying it here so that you can also leverage it if require. I think it can be highly helpful as a template to write MWEs for other yet-to-spot BUGs. 

```julia
using LinearAlgebra
using SparseArrays
using Gridap
using GridapGeosciences
using GridapHybrid


function Gridap.Geometry.push_normal(invJt,n)
  v = invJt⋅n
  m = sqrt(inner(v,v))
  if m < eps()
    return zero(v)
  else
    return v/m
  end
end

function project_initial_conditions(dΩ, P, Q, p₀, U, V, u₀, mass_matrix_solver)
  # the tracer
  b₁(q)    = ∫(q*p₀)dΩ
  a₁(p,q)  = ∫(q*p)dΩ
  rhs₁     = assemble_vector(b₁, Q)
  L2MM     = assemble_matrix(a₁, P, Q)
  L2MMchol = numerical_setup(symbolic_setup(mass_matrix_solver,L2MM),L2MM)
  pn       = FEFunction(Q, copy(rhs₁))
  pnv      = get_free_dof_values(pn)

  solve!(pnv, L2MMchol, pnv)

  # the velocity field
  b₂(v)    = ∫(v⋅u₀)dΩ
  a₂(u,v)  = ∫(v⋅u)dΩ
  rhs₂     = assemble_vector(b₂, V)
  MM       = assemble_matrix(a₂, U, V)
  MMchol   = numerical_setup(symbolic_setup(mass_matrix_solver,MM),MM)
  un       = FEFunction(V, copy(rhs₂))
  unv      = get_free_dof_values(un)

  solve!(unv, MMchol, unv)

  pn, pnv, L2MM, un
end

order  = 1
degree = 3
n      = 16
dt     = 0.25*2.0*π/((order+1)*4*n)
# single rotation about the sphere
nstep  = Int(2.0*π/dt)

# solid body rotation velocity field
function u₀(xyz)
  θϕr = xyz2θϕr(xyz)
  u   = cos(θϕr[2])
  spherical_to_cartesian_matrix(θϕr)⋅VectorValue(u,0,0)
end

# Gaussian tracer initial condition
function p₀(xyz)
  rsq = (xyz[1] - 1.0)*(xyz[1] - 1.0) + xyz[2]*xyz[2] + xyz[3]*xyz[3]
  exp(-4.0*rsq)
end

model = CubedSphereDiscreteModel(1,2; radius=1)

# Forward integration of the advection equation
D = num_cell_dims(model)
Ω = Triangulation(ReferenceFE{D},model)
Γ = Triangulation(ReferenceFE{D-1},model)
∂K = GridapHybrid.Skeleton(model)

reffeᵤ = ReferenceFE(lagrangian,VectorValue{3,Float64},order;space=:P)
reffeₚ = ReferenceFE(lagrangian,Float64,order;space=:P)
reffeₗ = ReferenceFE(lagrangian,Float64,order;space=:P)

# Define test FESpaces
V = TestFESpace(Ω, reffeᵤ; conformity=:L2)
Q = TestFESpace(Ω, reffeₚ; conformity=:L2)
M = TestFESpace(Γ, reffeₗ; conformity=:L2)
Y = MultiFieldFESpace([Q,M])

U = TrialFESpace(V)
P = TrialFESpace(Q)
L = TrialFESpace(M)
X = MultiFieldFESpace([P,L])

dΩ  = Measure(Ω,degree)
d∂K = Measure(∂K,degree)

# HDG stabilisation parameter
τ = 1.0

# Project the initial conditions onto the trial spaces
pn, pnv, L2MM, un = project_initial_conditions(dΩ, P, Q, p₀, U, V, u₀, Gridap.Algebra.BackslashSolver())

dth = 0.5*dt

n  = get_cell_normal_vector(∂K)
nₒ = get_cell_owner_normal_vector(∂K)


# First stage
b₁((q,m)) = ∫(q*pn)dΩ - ∫(m*0.0)d∂K
a₁((p,l),(q,m)) = ∫(q*p - dt*(∇(q)⋅un)*p)dΩ + ∫(((un⋅n) + τ*(n⋅n))*dt*q*p)d∂K -   # [q,p] block
              ∫(dt*q*l*τ*(n⋅n))d∂K +                                              # [q,l] block
              ∫(((un⋅nₒ) + τ*(n⋅nₒ))*m*p)d∂K -                                    # [m,p] block
              ∫(m*l*τ*(n⋅nₒ))d∂K                                                  # [m,l] block


a11(p,q)=∫(q*p - dt*(∇(q)⋅un)*p)dΩ + ∫(((un⋅n) + τ*(n⋅n))*dt*q*p)d∂K
A11=assemble_matrix(a11,X[1],Y[1])
a12(l,q)=∫(-dt*q*l*τ*(n⋅n))d∂K
A12=assemble_matrix(a12,X[2],Y[1])
a21(p,m)=∫(((un⋅nₒ) + τ*(n⋅nₒ))*m*p)d∂K
A21=assemble_matrix(a21,X[1],Y[2])
a22(l,m)=∫(-1.0*m*l*τ*(n⋅nₒ))d∂K
A22=assemble_matrix(a22,X[2],Y[2])

b1(q)= ∫(q*pn)dΩ
b2(m)=∫(-m*0.0)d∂K

A=assemble_matrix(a₁,X,Y)

B1=assemble_vector(b1,Y[1])
B2=assemble_vector(b2,Y[2])

BΓ  = B2  - Matrix(A21)*(A11\B1)
BΓΓ = A22 - Matrix(A21)*inv(Matrix(A11))*Matrix(A12)

op₁  = HybridAffineFEOperator((u,v)->(a₁(u,v),b₁(v)), X, Y, [1], [2])
Xh   = solve(op₁)
ph,_ = Xh

op_BΓΓ = op₁.skeleton_op.op.matrix
op_BΓ = op₁.skeleton_op.op.vector

@assert norm(op_BΓΓ-BΓΓ) < 1.0e-12
norm(BΓ-op_BΓ) < 1.0e-12
```

davelee2804 commented:
Hey @amartinhuertas 

> Why u\cdot n_o integrated over both sides and then summed is (n^+ - n^-)\cdot n_o? In the latter expression, I do not see the u. ??? What am I missing ...

Sorry, silly typo, that should read `(u^+ - u^-)\cdot n_o`, silly typo on my part

> I think this equivalence is only approximate in this case, as the normals might be contained in different planes

Great point, yes, I think as one refines the mesh `n^+` _converges to_ `-n^-`, but it will probably not be exact (the H(div) Piola map gives this exactly though yes?)

davelee2804 committed:
[ci skip] fixing a bug in the hdg advection solver (still now working
however...)

davelee2804 committed:
[ci skip] using the regular normal in the skeleton equation and the cell
owner normal in the advection equation. results look more coherant, but
still oscillatory

amartinhuertas commented:
hi @davelee2804 ... I am trying to make sense out of the HDG formulation that you have written, but I am struggling quite a bit (I cannot say whether it is correct or not, just that I dont understand). 

Looking at https://arxiv.org/abs/1605.03228, Eqs. (3.1)-(3.4), they briefly present the HDG upwind formulation for the (steady-state) advection equation (originally proposed in ref [5] https://users.oden.utexas.edu/~tanbui/PublishedPapers/phoga.pdf). Is this the formulation that are you trying to implement? 

For convenience, the correspondence among the article notation and the code notation is:
 * `\beta -> un`
 * `u -> p`
 * `v -> q`
 * `\hat{u} -> l`
 * `\mu -> m`

By now, I have sevaral concerns:
 1. The article mentions that `beta` is assumed to be a continuous function. However, `un` being used in the code is the projection of a continuous function into a discontinuous FE space. Correct? I would like to understand why `beta` has to be continuous.
 2. The term which is integrated over the bulk (i.e., K) in the article does not match the one in the code. 
 3. I do not know how to proof the equivalence among the terms in the code and the terms in the article over \partial K.

Perhaps you are trying to implement a different formulation ... not sure ....


![Screenshot from 2022-07-05 09-50-21](https://user-images.githubusercontent.com/38347633/177226297-6a74b340-acf1-497b-930a-4d4ac2a0d516.png)

davelee2804 commented:
Hey @amartinhuertas , 

> Is this the formulation that are you trying to implement?

They are doing material form advection `\beta.\nabla u`, whereas I am doing flux form `\nabla\cdot(\beta u()` For a divergence free vector field, `\nabla\cdot\beta = 0`, the two are the same. The velocity field we are using `(\cos(\phi),0)` is divergence free in the continuous form, but for `\beta\in H1(\Omega` I would imagine there will be some numerical divergence, so they won't be exactly the same

> However, un being used in the code is the projection of a continuous function into a discontinuous FE space. Correct? I would like to understand why beta has to be continuous.

We could do `un` continuous or discrete. We are only solving a single equation `u -> p` so there is the velocity field `\beta->un` will always be prescribed. When we extend to the shallow water equations then we solve for `\beta->un`

> The term which is integrated over the bulk (i.e., K) in the article does not match the one in the code.

Indeed, that is a result of the marerial vs. flux form advection. 

I don't understand the differences in the mesh skeleton term w.r.t. Kang et al 2020. (ie the use of `|\bea\cdot n|`) - do you have any thoughts on this??

amartinhuertas commented:
> They are doing material form advection \beta.\nabla u, whereas I am doing flux form \nabla\cdot(\beta u) For a divergence free vector field, \nabla\cdot\beta = 0, the two are the same.

Ok, I now understand that if the fluid flow is incompressible, then both the term in the code and the one in the article are equivalent, as I have double checked in the following note. Agreed?

![Screenshot from 2022-07-06 21-27-11](https://user-images.githubusercontent.com/38347633/177540840-0d54c884-69db-40fb-aeb2-27782469a793.png)

amartinhuertas commented:
> We could do un continuous or discrete. 

I would do it continuous (as far as possible). Is there any reason you need to interpolate it into a FE space? The only concern is that if we have a differential operator applied to the continuous solution we have to ensure that we apply the spherical variant of the differential operator, as we do, e.g., in the Darcy test with the laplacian applied to the analytical solution.

> We are only solving a single equation u -> p so there is the velocity field \beta->un will always be prescribed. When we extend to the shallow water equations then we solve for \beta->un

Yes, I was aware of this. \beta is input data to the advection problem.

amartinhuertas commented:
> I don't understand the differences in the mesh skeleton term w.r.t. Kang et al 2020. (ie the use of |\bea\cdot n|) - do you have any thoughts on this??

I think that Kang et al and the iHDG paper (the one from which I extracted the capture above) use the same abstract family of HDG formulations for hyperbolic problems of the general form

![Screenshot from 2022-07-06 21-45-36](https://user-images.githubusercontent.com/38347633/177543031-aece1151-fe95-47f6-947b-c55559b8a5cc.png)

For the advection problem (3.1), we have that 

```
F_k(u)=\beta_k u 
C=0
```

Thus, `A_k` is a 3x3 tensor with entries defined as `[A_k]_kk=\beta_k` and 0 otherwise. 

Agreed? (please let me know if everything clear)

amartinhuertas commented:
The general form of the upwind HDG discretization for problem (2.1) reads as:

![Screenshot from 2022-07-06 21-51-22](https://user-images.githubusercontent.com/38347633/177543943-9ee29a4d-5cc2-470a-8ef4-e1a1f6164782.png)

In the case of problem (3.1), A is diagonal matrix, with \beta_k n_k in the diagonal, for k=1..3. Thus, after computing the eigendecomposition of A, R is the identity, and S is equal to A, and |A| is |\beta \cdot n|, which is precisely what they use in (3.2) and (3.3). This choice of the upwind flux is the solution of a Riemman problem on the interface of the cells (TBH, I still do not fully understand this part, I need to devote more time to understand where the definition of the \hat{F} flux comes from and why it provides the desired convergence, stability, etc. properties.).

I guess that in Kang et. al. they are leveraging the eigenvalues of the A matrix that appears for the shallow water equations, and thus this does not apply verbatim to the (much simpler) advection equation.

How did you derive the formulation in the code for the advection equation from the one  of the shallow water equations in Kang et. al.? Could this be what is wrong?

davelee2804 commented:
Hey @amartinhuertas , I think I agree with what you say. If I unroll the definition of `\hat{F}\cdot n` from eqn (3) in Muralikrishnan et al as used in section (4) of that paper I get:

`\hat{F}\cdot n = \beta\cdot n u + |\beta\cdot n|(u - \hat{u})`, as in eqn (19), where `\beta` is the trial eigenvalue for the advection problem

if I unroll (17) from Kang et al (which is what I was trying to do) I get (using the same nomenclature) the same (using `n\cdot n = 1`)

Based on this I think my current implementation is wrong in two ways:
1. I am not using an absolute value, and 
2. I am not using the correct `\tau` (I am just using `\tau=1` for convenience, based on the max value of the absolute value of `|\beta\cdot n|` for this simple problem.

Is there a way to apply an absolute value in Gridap? I know there is `mean()` and `jump()`, is there also an `abs()` ?

amartinhuertas commented:
> I know there is mean() and jump(), is there also an abs() ?

I think that is possible. AFAIK, the high level API supports an invokation to an arbitrary Julia function passing a `CellField` as an argument (\beta \cdot n is a `CellField`). Anyway, I have never tried that with `GridapHybrid.jl` ... let me know if that does not work.

amartinhuertas commented:
BTW, did you notice that in (2.2a) they are using exactly the same bulk term as the one you are using in your code? :-)

davelee2804 committed:
mass is conserved now and results are non-oscillatory. just need to
figure out how to implement the jump() operator...

davelee2804 commented:
> BTW, did you notice that in (2.2a) they are using exactly the same bulk term as the one you are using in your code? :-)

I haven't checked it experimentally, but for a divergence free `\beta` I think it should be OK. 

I've just made a change that leads to a much more coherent solution (and exact mass conservation), however, there is no `jump()` in there now as this leads to a code error as: `It is not possible to perform the operation "*" on the given cell fields.`

Do you know if the `jump()` operator is supported for integration with respect to test functions on the Skeleton using the measure `d\partial K` ?

amartinhuertas commented:
> Do you know if the jump() operator is supported for integration with respect to test functions on the Skeleton using the measure d\partial K ?

Yes. The `jump()` operator is not supported, on purpose. The measure d\partial K represents the boundary of the cells, on a cell by cell basis, such that each cell is independent of each other. The `jump()` operator can be used only with facet-wise triangulations that glue together the two cells around. I am not sure why you need the jump() operator explicity.

davelee2804 commented:
OK, great to know, thanks for the confirmation @amartinhuertas . Well, how do we represent `[[.]]` then? I guess this is what `n\_o` is for... I'll keep trying to find a way to replicate the jump() with this. Right now the results look coherant and conservative, but I think over long times they may go oscillatory without some sort of jump penalisation...

amartinhuertas commented:
> Well, how do we represent [[.]] then?

Let us assume that we want to compute, e.g., 

`< [[u]], v>_e`, for all edges `e` in the mesh, where `[[u]] = u^+ - u^-` 

Then you have to write 

`< u*(n \cdot n_o), v>_d\partial K`, for all cells `K` in the mesh.

So I think that you already got that from what I see in your comments.

amartinhuertas commented:
> Right now the results look coherant and conservative, but I think over long times they may go oscillatory without some sort of jump penalisation...

This is great! I am surprised that we only needed a moderate amount of BUG fixes. :-)

amartinhuertas commented:
> I haven't checked it experimentally, but for a divergence free \beta I think it should be OK.

I am surprised they do not mention that \beta is assumed to be divergence free at any point in the iHDG paper ... Perhaps it is something well-known in hyperbolic fluid flow problems or something like that ...

davelee2804 commented:
> This is great! I am surprised that we only needed a moderate amount of BUG fixes. :-)

Doesn't mean it is working correctly yet however :). Could be that it will still go oscillatory or not advect correctly (I only checked a couple of time steps). 

When I add in the jumps as `n\cdot n\_o` I get completely nonsense results (and a total loss of mass conservation), so I am wondering if there is some subtle interplay between the `n\cdot n_o` jump and the `|A|` absolute value that I don't properly understand....

davelee2804 commented:
> I am surprised they do not mention that \beta is assumed to be divergence free at any point in the iHDG paper ... Perhaps it is something well-known in hyperbolic fluid flow problems or something like that ...

`\beta` could have a divergence. It is just that if `\beta` has non-zero divergence then the material and flux form advection equations are no longer equivalent. I thought I would do flux form as that is how the advection equation is done in shallow water.

amartinhuertas commented:
> \beta could have a divergence. It is just that if \beta has non-zero divergence then the material and flux form advection equations are no longer equivalent. I thought I would do flux form as that is how the advection equation is done in shallow water.

Ok. Thanks for the clarification. I think I 100% understand now.

amartinhuertas commented:
> Doesn't mean it is working correctly yet however :). Could be that it will still go oscillatory or not advect correctly (I only checked a couple of time steps).
> When I add in the jumps as n\cdot n\_o I get completely nonsense results (and a total loss of mass conservation), so I am wondering if there is some subtle interplay between the n\cdot n_o jump and the |A| absolute value that I don't properly understand..

To which kind of oscillations are you referring to? You mean spurious modes that you usually have in the Galerkin method if you do not use upwinding? (such as SUPG?). 

I am not wrong, the formulation we are leveraging already comes with upwinding "batteries-on", no need to do anything else. But perhaps I am missing something ...

davelee2804 commented:
Yes, the formulation we are trying to implement includes upwinding, agreed. However I am not able to get the jump penalisation working, which is why I think it will eventually go oscillatory...

davelee2804 commented:
...Does the integral over the measure `d\partial K` already assume that the jump is baked in?

amartinhuertas commented:
> I am not able to get the jump penalisation working

Which is the evidence that supports this statement?

amartinhuertas commented:
> ...Does the integral over the measure d\partial K already assume that the jump is baked in?

No, AFAIK.

amartinhuertas commented:
Have you noticed that both Kang et al and MURALIKRISHNAN et al defined the jump as [[u]] = u^+ + u^- ??? With a + not a -

amartinhuertas commented:
> Which is the evidence that supports this statement?

Ok, I think you already answered to that, sorry. Namely here

> ie: the n\cdot n\_o term leads to large errors...

I am not sure we have to use n\cdot n\_o here, given the definition of the jump operator in that papers.

davelee2804 committed:
[ci skip] very minor changes to the model configuration

davelee2804 commented:
Hey @amartinhuertas , as mentioned, the application of 'n\cdot n_o` leads to erroneous results for me. With the current configuration the results are correct (conservative and advecting at the correct speed), see images attached. So I'm very confused about the application of the jumps. In this article (which also discusses continuity of the tangent vectors), there are no jumps explicitly mentioned (see eqn 7 for example): https://www.sciencedirect.com/science/article/pii/S0021999111003226



![time_0T](https://user-images.githubusercontent.com/11448075/177938631-6d93bb59-01a9-4240-87bc-aaac090a6135.png)
![time_1T](https://user-images.githubusercontent.com/11448075/177938642-356d3a59-06b9-40f3-9baa-8900fd8e5fbb.png)
![time_Ton2](https://user-images.githubusercontent.com/11448075/177938645-e8b93e17-2ada-426a-8ad6-1e33d7972aab.png)
![time_Ton4](https://user-images.githubusercontent.com/11448075/177938650-62747ed8-6395-4fbb-b5fa-7fa8d7185d11.png)
:

davelee2804 commented:
...so I have no idea if the penalisation is properly applied or not. Given that the results are non-oscillatory and conservative, it looks like _maybe_ this is being done correctly, but I don't know enough about how this works under the hood to say. Also I will need to implement a more sophisticated version of `\tau` (as `|u\cdot n|`, not `1`).

amartinhuertas commented:
> In this article (which also discusses continuity of the tangent vectors), there are no jumps explicitly mentioned (see eqn 7 for example): https://www.sciencedirect.com/science/article/pii/S0021999111003226

Hi @davelee2804 ! The jumps are defined here, yes? (Again, as in the other papers, with a + not a -, i.e., no need for n \cdot no)

![Screenshot from 2022-07-08 17-45-09](https://user-images.githubusercontent.com/38347633/177943333-aaf1b966-4fdb-4f5f-bf09-a0ddc60ca0b2.png)

amartinhuertas commented:
> Also I will need to implement a more sophisticated version of \tau (as |u\cdot n|, not 1).

In my experience, if \tau is not set to the appropriate value, one of the effects might be that the method does not convergence at the proper order as you refine the mesh.

davelee2804 committed:
[ci skip] implemented the correct form of \tau (|u.n|), results are
marginally improved

amartinhuertas commented:
Note that an integral over \partial K is nothing but a sum of integrals over the edges of K with the integrand restricted to each edge seen from the perspective of K.

davelee2804 commented:
Thank you @amartinhuertas ! I had not appreciated the subtlety that in for example Kang et al. eqns (14) and (15) the first skeleton integral is over `\partial K` while the second is over `e`, this makes much more sense now! Looking at (7) in Nguyen et al I see that all the skeleton integrals are over `\partial T`, which is consistent with 
 
> integrand restricted to each edge seen from the perspective of K

as you say. 

I am now implementing a convergence study for the solid body rotation test (second order in space + time), to verify that everything is OK.

A few days ago I got some nonsensical results with order 4, so I will also double check that as well...

davelee2804 committed:
[ci skip] computing the l2 error at the end of the revolution

amartinhuertas commented:
> Thank you @amartinhuertas ! I had not appreciated the subtlety that in for example Kang et al. eqns (14) and (15) the first skeleton integral is over \partial K while the second is over e, this makes much more sense now! Looking at (7) in Nguyen et al I see that all the skeleton integrals are over \partial T, which is consistent with

Exactly. You can write hybridizable methods either by edges or grouping together edges into cell boundaries \partial K, for each K. The second approach is the one required to implement static condensation locally at each cell, towards assembly of the global problem on the skeleton, which is the main motivation for hybridizable methods.

> I am now implementing a convergence study for the solid body rotation test (second order in space + time), to verify that everything is OK.
> A few days ago I got some nonsensical results with order 4, so I will also double check that as well...

Ok. Let us see.

davelee2804 commented:
Hey @amartinhuertas , just a heads up that the convergence is sub-optimal. There is too much dispersion, perhaps as a result of how I have formulated the time stepping. I'll take a closer look at this...

amartinhuertas commented:
>  Hey @amartinhuertas , just a heads up that the convergence is sub-optimal. There is too much dispersion, perhaps as a result of how I have formulated the time stepping. I'll take a closer look at this...

I would try to first solve a steady advection problem with manufactured PDE solution. This way we eliminate noise related to time discretization.

davelee2804 committed:
[ci skip] using stiffly stable second order implicit runge kutta.
also removing the (n.n) terms. results greatly improved

davelee2804 commented:
Thanks @amartinhuertas , yes great point, this is the correct approach, I will do this. 

Just a heads up, I changed the time integration to a stiffly-stable second order implicit runge-kutta scheme, and now the solid body rotation test is converging at the correct rate (second order convergence for second order in space+time discretisation) after a single revolution round the sphere (at least for the low resolutions that I have checked).

davelee2804 commented:
![hdg_convergence](https://user-images.githubusercontent.com/11448075/178144757-7c092ed8-2ef9-45d9-9bff-e95b88faff3d.png)

davelee2804 committed:
adding convergence test for the HDG advection solver to the ci test
suite

davelee2804 commented:
Hi @amartinhuertas , I've added a convergence test for the solid body rotation configuration of the HDG advection solver. I'm wondering if now is a good time to do a code review and merge this branch, before starting work on the linear wave equation with jump penalisation on the pressure gradient tangent component - what do you think?

davelee2804 commented:
...SBRAdvectionHDG tests are failing in github-ci (but passing on my laptop) :
https://github.com/gridapapps/GridapGeosciences.jl/runs/7274708478?check_suite_focus=true#step:9:1037

Should I also commit the Manifest.toml?

amartinhuertas commented:
> Should I also commit the Manifest.toml?

Yes, because we are pointing to versions of some of the packages which are not yet registered in the Julia registries.

amartinhuertas commented:
Also, please, pull from `master` if you didnt lately. I see in this PR the following changes that are already in `master`, so that they should not be in this PR. 


https://github.com/gridapapps/GridapGeosciences.jl/pull/34/files#diff-8ad2a593c8bedda9e0bea392ea69c0bbbe41609839d4d3b33bc4a81155a208dcL48

davelee2804 committed:
[ci skip] committing the manifest

davelee2804 committed:
Merge branch 'master' of github.com:gridapapps/GridapGeosciences.jl into hdg_adv

amartinhuertas commented:
> Hi @amartinhuertas , I've added a convergence test for the solid body rotation configuration of the HDG advection solver. 

I see in the attached plot that there is a mild degradation in slope for the highest resolution that you tested. Have you tried even higher resolutions to be 100% sure that the convergence is actually quadratic?

davelee2804 commented:
> I see in the attached plot that there is a mild degradation in slope for the highest resolution that you tested

Hi @amartinhuertas , higher resolutions are to the left on the x-axis, so the convergence rate is actually increasing as resolution increases. Does that make sense? Have I missed your point somehow?

davelee2804 commented:
...Ie: convergence rate from dx=0.12 to 0.06 is higher than from dx=0.24 to 0.12. Does that make sense?

amartinhuertas commented:
> ...Ie: convergence rate from dx=0.12 to 0.06 is higher than from dx=0.24 to 0.12. Does that make sense?

Ok, you are right, I was confused. The convergence curve  looks like a piece-wise linear function with 2 pieces, where the right piece seems to be parallel to the theoretical convergence, and the left piece seems to bend down a little bit, with an slope which is no longer parallel to the theoretical convergence. But, as you say, bending down means even faster convergence, and not the opposite. (Here it was my confusion).

amartinhuertas commented:
> I'm wondering if now is a good time to do a code review and merge this branch

Yes, it might be a good time to do this. The only thing is that I have some other code reviews on the queue, I wont be able to do this immediately.

Apart from this code review, we need:
1) To investigate issue https://github.com/gridapapps/GridapGeosciences.jl/issues/36
2) To fix (if actually required) and test the geometrical discretization of the sphere using an analytical mapping.

davelee2804 commented:
No problem. Note that for the SBRAdvectionHDGTests file that I committed, convergence between the lowest two resolutions is sub-optimal, but quadratic convergence is achieved between the next two.

davelee2804 commented:
No problem @amartinhuertas , there is no rush from me. Please let me know if there is anything I can do to help with any of the issues you mention. We are using an isoparametric mapping right now, is that correct?

amartinhuertas commented:
> We are using an isoparametric mapping right now, is that correct?

We are using biquadratic elements for the geometry, and bilinear elements for the unknown. I think that the term "isoparametric" is used to refer to those scenarios in which the order of the polynomial to approximate the geometry matches the order of the polynomial to approximate the unknown.

davelee2804 commented:
Aaah, yes, that is what the `2` is for in the CubedSphereDiscreteModel :) OK. Is there a particular reason we need to support an analytical mapping as well?

amartinhuertas commented:
> Is there a particular reason we need to support an analytical mapping as well?

We have been using an analytical mapping in all the results we obtained with GridapGeosciences + compatible FEs so far. 

If we can represent the geometry exactly, better to stick to it, instead of approximating it, so that we eliminate any noise caused by geometrical errors (leaving integration errors apart). I remember e.g., that, with a bilinear approximation of the cubed sphere we were not able to obtained the theoretical convergence rates of RT+DG FEs for Darcy on the sphere.

davelee2804 commented:
OK, fair enough. To my mind supporting iso or super parametric mappings (as you are already doing) is better, as if we ever want to add mountains when we will have to ditch the analytic mapping anyway, so ensuring everything works for a numerical mapping makes it easier to transition to such a case if we ever want to. But I take your point, analytic is nicer...

amartinhuertas commented:
Yes, sure, better to have both.

amartinhuertas committed:
Merge branch 'master' of github.com:gridapapps/GridapGeosciences.jl into hdg_adv

davelee2804 committed:
checking the hdg advection tests

davelee2804 committed:
Merge branch 'hdg_adv' of github.com:gridapapps/GridapGeosciences.jl into hdg_adv
